Record with "is" prefixed Boolean field causes unmapped source property report

Given files:

#### MemberDto.java

```java
package org.mapstruct.itest.records;

public record MemberDto(Boolean isActive, Boolean premium) {

}
```

#### MemberEntity.java

```java
package org.mapstruct.itest.records;

public class MemberEntity {

    private Boolean isActive;
    private Boolean premium;

    public Boolean getIsActive() {
        return isActive;
    }

    public void setIsActive(Boolean isActive) {
...
```

#### MemberMapper.java

```java
package org.mapstruct.itest.records;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.ReportingPolicy;
import org.mapstruct.factory.Mappers;

@Mapper(unmappedSourcePolicy = ReportingPolicy.ERROR)
public interface MemberMapper {

    MemberMapper INSTANCE = Mappers.getMapper( MemberMapper.class );

    MemberEntity fromRecord(MemberDto record);

    MemberDto toRecord(MemberEntity entity);

}
```

And following method added to org.mapstruct.itest.records.RecordsTest:

```java
import org.mapstruct.itest.records.MemberDto;
import org.mapstruct.itest.records.MemberEntity;
import org.mapstruct.itest.records.MemberMapper;

...

	@Test
    public void shouldMapMemberRecord() {
        MemberEntity member = MemberMapper.INSTANCE.fromRecord( new MemberDto( true, false ) );

        assertThat( member ).isNotNull();
        assertThat( member.getIsActive() ).isTrue();
        assertThat( member.getPremium() ).isFalse();
    }

    @Test
    public void shouldMapIntoMemberRecord() {
        MemberEntity entity = new MemberEntity();
        entity.setIsActive( false );
        entity.setPremium( true );

        MemberDto value = MemberMapper.INSTANCE.toRecord( entity );

        assertThat( value ).isNotNull();
        assertThat( value.isActive() ).isEqualTo( false );
        assertThat( value.premium() ).isEqualTo( true );
    }
```

Test failed with compilation error:

> [ERROR] /.../mapstruct/integrationtest/target/tmp/org.mapstruct.itest.tests.recordsTest/javac/src/main/java/org/mapstruct/itest/records/MemberMapper.java:[22,18] Unmapped source property: "active".

(But actually I cannot find any warning when removing `unmappedSourcePolicy`, maybe the maven test task just remove warnings?)

### Environment

MapStruct 1.5.0-SNAPSHOT (as of 8b84f5b7d78979be979940b8ded1cbbeb868e32c)

#### Java

> openjdk version "17" 2021-09-14 LTS
> OpenJDK Runtime Environment Zulu17.28+13-CA (build 17+35-LTS)
> OpenJDK 64-Bit Server VM Zulu17.28+13-CA (build 17+35-LTS, mixed mode, sharing)

--- a/integrationtest/src/test/resources/recordsTest/src/main/java/org/mapstruct/itest/records/MemberDto.java
+++ b/integrationtest/src/test/resources/recordsTest/src/main/java/org/mapstruct/itest/records/MemberDto.java
@@ -1,0 +1,13 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.itest.records;
+
+/**
+ * @author Filip Hrisafov
+ */
+public record MemberDto(Boolean isActive, Boolean premium) {
+
+}

--- a/integrationtest/src/test/resources/recordsTest/src/main/java/org/mapstruct/itest/records/MemberEntity.java
+++ b/integrationtest/src/test/resources/recordsTest/src/main/java/org/mapstruct/itest/records/MemberEntity.java
@@ -1,0 +1,31 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.itest.records;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class MemberEntity {
+
+    private Boolean isActive;
+    private Boolean premium;
+
+    public Boolean getIsActive() {
+        return isActive;
+    }
+
+    public void setIsActive(Boolean active) {
+        isActive = active;
+    }
+
+    public Boolean getPremium() {
+        return premium;
+    }
+
+    public void setPremium(Boolean premium) {
+        this.premium = premium;
+    }
+}

--- a/integrationtest/src/test/resources/recordsTest/src/main/java/org/mapstruct/itest/records/MemberMapper.java
+++ b/integrationtest/src/test/resources/recordsTest/src/main/java/org/mapstruct/itest/records/MemberMapper.java
@@ -1,0 +1,24 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.itest.records;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.ReportingPolicy;
+import org.mapstruct.factory.Mappers;
+
+/**
+ * @author Filip Hrisafov
+ */
+@Mapper(unmappedSourcePolicy = ReportingPolicy.ERROR)
+public interface MemberMapper {
+
+    MemberMapper INSTANCE = Mappers.getMapper( MemberMapper.class );
+
+    MemberEntity fromRecord(MemberDto record);
+
+    MemberDto toRecord(MemberEntity entity);
+
+}

--- a/integrationtest/src/test/resources/recordsTest/src/test/java/org/mapstruct/itest/records/RecordsTest.java
+++ b/integrationtest/src/test/resources/recordsTest/src/test/java/org/mapstruct/itest/records/RecordsTest.java
@@ -61,4 +61,26 @@ public class RecordsTest {
         assertThat( carDto.wheelPositions() )
             .containsExactly( "left" );
     }
+
+    @Test
+    public void shouldMapMemberRecord() {
+        MemberEntity member = MemberMapper.INSTANCE.fromRecord( new MemberDto( true, false ) );
+
+        assertThat( member ).isNotNull();
+        assertThat( member.getIsActive() ).isTrue();
+        assertThat( member.getPremium() ).isFalse();
+    }
+
+    @Test
+    public void shouldMapIntoMemberRecord() {
+        MemberEntity entity = new MemberEntity();
+        entity.setIsActive( false );
+        entity.setPremium( true );
+
+        MemberDto value = MemberMapper.INSTANCE.toRecord( entity );
+
+        assertThat( value ).isNotNull();
+        assertThat( value.isActive() ).isEqualTo( false );
+        assertThat( value.premium() ).isEqualTo( true );
+    }
 }

--- a/processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java
@@ -573,14 +573,33 @@ public class Type extends ModelElement implements Comparable<Type> {
      */
     public Map<String, Accessor> getPropertyReadAccessors() {
         if ( readAccessors == null ) {
-            List<Accessor> getterList = filters.getterMethodsIn( getAllMethods() );
             Map<String, Accessor> modifiableGetters = new LinkedHashMap<>();
+
+            Map<String, Accessor> recordAccessors = filters.recordAccessorsIn( getRecordComponents() );
+            modifiableGetters.putAll( recordAccessors );
+
+            List<Accessor> getterList = filters.getterMethodsIn( getAllMethods() );
             for ( Accessor getter : getterList ) {
+                String simpleName = getter.getSimpleName();
+                if ( recordAccessors.containsKey( simpleName ) ) {
+                    // If there is already a record accessor that contains the simple name
+                    // then it means that the getter is actually a record component.
+                    // In that case we need to ignore it.
+                    // e.g. record component named isActive.
+                    // The DefaultAccessorNamingStrategy will return active as property name,
+                    // but the property name is isActive, since it is a record
+                    continue;
+                }
                 String propertyName = getPropertyName( getter );
+
+                if ( recordAccessors.containsKey( propertyName ) ) {
+                    // If there is already a record accessor, the property needs to be ignored
+                    continue;
+                }
                 if ( modifiableGetters.containsKey( propertyName ) ) {
                     // In the DefaultAccessorNamingStrategy, this can only be the case for Booleans: isFoo() and
                     // getFoo(); The latter is preferred.
-                    if ( !getter.getSimpleName().startsWith( "is" ) ) {
+                    if ( !simpleName.startsWith( "is" ) ) {
                         modifiableGetters.put( propertyName, getter );
                     }
 
@@ -588,11 +607,6 @@ public class Type extends ModelElement implements Comparable<Type> {
                 else {
                     modifiableGetters.put( propertyName, getter );
                 }
-            }
-
-            Map<String, Accessor> recordAccessors = filters.recordAccessorsIn( getRecordComponents() );
-            for ( Map.Entry<String, Accessor> recordEntry : recordAccessors.entrySet() ) {
-                modifiableGetters.putIfAbsent( recordEntry.getKey(), recordEntry.getValue() );
             }
 
             List<Accessor> fieldsList = filters.fieldsIn( getAllFields() );

